src/FileCacheManager.php

```diff
@@ -32,159 +37,100 @@
+        $this->cacheDirectory = $cacheDirectory ?: new Directory(dirname(realpath($handler->getFile())));
+
+        $this->readCache();
     }
 
     public function __destruct()
     {
-        $this->saveToFile();
+        $this->writeCache();
     }
 
-    public function needFixing($file, $fileContent)
+    private function readCache()
     {
-        if (!$this->isCacheAvailable()) {
-            return true;
-        }
+        $cache = $this->handler->read();
 
-        $file = $this->getRelativePathname($file);
-
-        if (!isset($this->oldHashes[$file])) {
-            return true;
-        }
-
-        if ($this->oldHashes[$file] !== $this->calcHash($fileContent)) {
-            return true;
+        if (!$cache || !$this->signature->equals($cache->getSignature())) {
+            $cache = new Cache($this->signature);
         }
 
-        // file do not change - keep hash in new collection
-        $this->newHashes[$file] = $this->oldHashes[$file];
-
-        return false;
+        $this->cache = $cache;
     }
 
-    public function setFile($file, $fileContent)
+    private function writeCache()
     {
-        if (!$this->isCacheAvailable()) {
-            return;
-        }
-
-        $file = $this->getRelativePathname($file);
-
-        $this->newHashes[$file] = $this->calcHash($fileContent);
+        $this->handler->write($this->cache);
     }
 
-    private function calcHash($content)
-    {
-        return crc32($content);
-    }
-
-    private function isCacheAvailable()
+    public function needFixing($file, $fileContent)
     {
-        static $result;
-
-        if (null === $result) {
-            $result = $this->isEnabled && (ToolInfo::isInstalledAsPhar() || ToolInfo::isInstalledByComposer());
-        }
+        $file = $this->cacheDirectory->relativePathTo($file);
 
-        return $result;
-    }
-
-    private function isCacheStale($php, $version, $linting, $rules)
-    {
-        if (!$this->isCacheAvailable()) {
+        if (!$this->cache->has($file) || $this->cache->get($file) !== $this->calcHash($fileContent)) {
             return true;
         }
 
-        return PHP_VERSION !== $php || ToolInfo::getVersion() !== $version || $this->linting !== $linting || $this->rules !== $rules;
+        return false;
     }
 
-    private function readFromFile()
+    public function setFile($file, $fileContent)
     {
-        if (!$this->isCacheAvailable()) {
-            return;
-        }
-
-        if (!file_exists($this->cacheFile)) {
-            return;
-        }
-
-        $content = file_get_contents($this->cacheFile);
-        $data = @unserialize($content);
-
-        // ignore corrupted serialized data
-        if (null === $data) {
-            return;
-        }
+        $file = $this->cacheDirectory->relativePathTo($file);
 
-        if (!isset($data['php'], $data['version'], $data['linting'], $data['rules'])) {
-            return;
-        }
+        $hash = $this->calcHash($fileContent);
```

Comment:
this is slow operation.
please, perform it only when needed.

for example, we shouldn't need it when we are using `NullHandler`.
But so far, even with `NullHandler` we are using `Cache` and calculate hash here, in manager. Maybe moving it to `CacheInterface` would allow us to move that logic out of this class and introduce `NullCache` ?